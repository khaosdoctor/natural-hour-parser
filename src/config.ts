import { z } from 'zod'
import debug from 'debug'
const baseLogger = debug('wolt')

baseLogger.extend('config')('Loading configuration')
export const AppConfigSchema = z.object({
	PORT: z
		.string()
		.optional()
		.default('3000')
		.transform((value) => Number(value))
		.refine((port) => port >= 0 && port <= 65535, {
			message: 'PORT must be between 0 and 65535',
		}),
})

export type AppConfig = z.infer<typeof AppConfigSchema> & {
	logger: typeof baseLogger
}
export function appConfig(origin = process.env): AppConfig {
	return {
		...AppConfigSchema.parse(origin),
		logger: baseLogger,
	}
}
