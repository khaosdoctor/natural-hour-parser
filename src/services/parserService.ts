import { type Debugger } from 'debug'
import { type AppConfig } from '../config.js'
import { HourParser } from '../domain/Parser.js'
import { type OpeningHoursInput } from '../domain/validation.js'

export class ParserService {
	readonly #logger: Debugger
	readonly #parser: HourParser
	constructor(config: AppConfig) {
		this.#logger = config.logger.extend(ParserService.name)
		this.#logger(`${ParserService.name} initialized`)
		this.#parser = new HourParser(config)
	}

	async parseOpeningHours(input: OpeningHoursInput) {
		this.#logger('Parsing opening hours')
		return this.#parser.parseOpeningHours(input)
	}

	formatIntervals(
		intervals: Awaited<ReturnType<HourParser['parseOpeningHours']>>
	) {
		this.#logger('Formatting intervals')
		return this.#parser.toText(intervals)
	}
}
