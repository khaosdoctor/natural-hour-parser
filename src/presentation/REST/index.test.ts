import debug from 'debug'
import request from 'supertest'
import { type AppConfig } from '../../config.js'
import { ParserService } from '../../services/parserService.js'
import { RESTInterface } from './index.js'

// Mock config
const mockConfig: AppConfig = {
	logger: debug('test:integration'),
	PORT: 3000,
}

const services = {
	parserService: new ParserService(mockConfig),
}

const input = {
	monday: [],
	tuesday: [
		{
			type: 'open',
			value: 36000,
		},
		{
			type: 'close',
			value: 64800,
		},
	],
	wednesday: [],
	thursday: [
		{
			type: 'open',
			value: 37800,
		},
		{
			type: 'close',
			value: 64800,
		},
	],
	friday: [
		{
			type: 'open',
			value: 36000,
		},
	],
	saturday: [
		{
			type: 'close',
			value: 3600,
		},
		{
			type: 'open',
			value: 36000,
		},
	],
	sunday: [
		{
			type: 'close',
			value: 3600,
		},
		{
			type: 'open',
			value: 43200,
		},
		{
			type: 'close',
			value: 75600,
		},
	],
}

const { app } = RESTInterface(mockConfig, services)
const wrappedApp = request(app)
describe('Integration Tests', () => {
	describe('GET /ping', () => {
		it('should return "pong"', async () => {
			const response = await wrappedApp.get('/ping')
			expect(response.statusCode).toBe(200)
			expect(response.text).toBe('pong')
		})
	})

	describe('POST /parsers/restaurants/opening-hours', () => {
		it('should return parsed opening hours in JSON format', async () => {
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send(input)
				.set('Accept', 'application/json')

			expect(response.status).toBe(200)
			expect(response.body).toMatchSnapshot()
			expect(response.header['content-type']).toContain('application/json')
		})

		it('should return parsed opening hours in text format', async () => {
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send(input)
				.set('Accept', 'text/plain')

			expect(response.status).toBe(200)
			expect(response.text).toMatchSnapshot()
			expect(response.header['content-type']).toContain('text/plain')
		})

		it('should return 406 on invalid formats', async () => {
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send(input)
				.set('Accept', 'text/csv')

			expect(response.status).toBe(406)
		})

		it('should strip unknown keys and treat as valid input', async () => {
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send({ invalid: 'input' })

			expect(response.status).toBe(200)
			expect(response.body).toMatchInlineSnapshot(`{}`)
		})

		it('should handle input errors', async () => {
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send({
					monday: [
						{
							type: 'not a valid type',
							value: 36000,
						},
					],
				})

			expect(response.status).toBe(422)
			expect(response.body).toMatchSnapshot()
		})

		it('should handle any other errors as boundaries', async () => {
			const mockParserService = {
				parseOpeningHours: jest.fn(() => {
					throw new Error('Something went wrong')
				}),
			}
			const { app } = RESTInterface(mockConfig, {
				parserService: mockParserService as any,
			})
			const wrappedApp = request(app)
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send({})

			expect(response.status).toBe(500)
			expect(response.body).toMatchSnapshot()
		})

		it('should handle domain errors', async () => {
			const wrappedApp = request(app)
			const response = await wrappedApp
				.post('/parsers/restaurants/opening-hours')
				.send({
					monday: [
						{
							type: 'open',
							value: 3600,
						},
					],
				})

			expect(response.status).toBe(422)
			expect(response.body).toMatchSnapshot()
		})
	})
})
