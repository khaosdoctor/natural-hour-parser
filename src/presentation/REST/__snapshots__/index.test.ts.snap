// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests POST /parsers/restaurants/opening-hours should handle any other errors as boundaries 1`] = `
{
  "message": "Internal server error",
}
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should handle domain errors 1`] = `
{
  "code": "INVALID_INPUT",
  "message": "Number of opening and closing hours must match",
  "name": "InvalidInputError",
}
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should handle input errors 1`] = `
{
  "issues": [
    {
      "code": "invalid_enum_value",
      "message": "Invalid enum value. Expected 'open' | 'close', received 'not a valid type'",
      "options": [
        "open",
        "close",
      ],
      "path": [
        "monday",
        0,
        "type",
      ],
      "received": "not a valid type",
    },
  ],
  "name": "ZodError",
}
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should return parsed opening hours in JSON format 1`] = `
{
  "friday": [
    {
      "close": "1970-01-01T01:00:00.000Z",
      "open": "1970-01-01T10:00:00.000Z",
    },
  ],
  "monday": [],
  "saturday": [
    {
      "close": "1970-01-01T01:00:00.000Z",
      "open": "1970-01-01T10:00:00.000Z",
    },
  ],
  "sunday": [
    {
      "close": "1970-01-01T21:00:00.000Z",
      "open": "1970-01-01T12:00:00.000Z",
    },
  ],
  "thursday": [
    {
      "close": "1970-01-01T18:00:00.000Z",
      "open": "1970-01-01T10:30:00.000Z",
    },
  ],
  "tuesday": [
    {
      "close": "1970-01-01T18:00:00.000Z",
      "open": "1970-01-01T10:00:00.000Z",
    },
  ],
  "wednesday": [],
}
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should return parsed opening hours in text format 1`] = `
"Sunday: 12:00 PM - 9:00 PM
Monday: Closed
Tuesday: 10:00 AM - 6:00 PM
Wednesday: Closed
Thursday: 10:30 AM - 6:00 PM
Friday: 10:00 AM - 1:00 AM
Saturday: 10:00 AM - 1:00 AM"
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should strip unknown keys and keep known ones 1`] = `
"Sunday: Closed
Monday: 1:00 AM - 2:00 AM
Tuesday: Closed
Wednesday: Closed
Thursday: Closed
Friday: Closed
Saturday: Closed"
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should strip unknown keys and treat as empty input 1`] = `
{
  "friday": [],
  "monday": [],
  "saturday": [],
  "sunday": [],
  "thursday": [],
  "tuesday": [],
  "wednesday": [],
}
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should treat empty days as closed 1`] = `
"Sunday: Closed
Monday: Closed
Tuesday: Closed
Wednesday: Closed
Thursday: Closed
Friday: Closed
Saturday: Closed"
`;

exports[`Integration Tests POST /parsers/restaurants/opening-hours should treat empty days as closed 2`] = `
"Sunday: Closed
Monday: Closed
Tuesday: Closed
Wednesday: Closed
Thursday: Closed
Friday: Closed
Saturday: Closed"
`;
